### PR TITLE
made parameters consistent in cpp and python code; added missed aliases to the docs

### LINKED ISSUE #1014: Different default hyperparameters in sklearn API
According to the hyperparameter docs (https://lightgbm.readthedocs.io/en/latest/Parameters.html):

`min_data_in_leaf, default=20, type=int, alias=min_data_per_leaf , min_data`

Whereas in the scikit-learn API (https://lightgbm.readthedocs.io/en/latest/Python-API.html#lightgbm.LGBMClassifier):

`min_child_samples (int, optional (default=10)) – Minimum number of data need in a child(leaf).`

As you can see, the default value for the hyperparameter changes.

The same happens in `min_sum_hessian_in_leaf`: in the booster API the default is `1e-3`, but in scikit-learn's API it is `5` for `min_child_weight`

Is there any reason for this change in the default value for those hyperparameters?

### PR DESCRIPTION
**Pull request by StrikerRUS**
fixed #1014.

I can't find description of these parameters. It seems that they haven't been included into Parameters.rst.

https://github.com/Microsoft/LightGBM/blob/b9dc51a6c58d98aeebf1b23c10c249c813e6e9fb/include/LightGBM/config.h#L363

https://github.com/Microsoft/LightGBM/blob/b9dc51a6c58d98aeebf1b23c10c249c813e6e9fb/include/LightGBM/config.h#L435

https://github.com/Microsoft/LightGBM/blob/b9dc51a6c58d98aeebf1b23c10c249c813e6e9fb/include/LightGBM/config.h#L442-L443

### CODE REVIEW

**Comment by wxchan** [comment]
I think it is kinda breaking back compatibility, if user didn't set this parameter, he may get different results after this pr. Maybe @ClimbsRocks can give some suggestions here?

**Comment by StrikerRUS** [comment]
@wxchan I suppose it's OK, since we don't break the existing users' codebase, but just change the behavior of the program and it's normal thing for developing process.

**Comment by guolinke** [comment]
@StrikerRUS can we let them None ?

**Comment by StrikerRUS** [comment]
@guolinke I'm afraid, I didn't understand you.

All these parameters have their certain default values, but they are not documented in Parameters.rst. I'm not talking about sklearn, it's all about `config.h`

https://github.com/Microsoft/LightGBM/blob/b9dc51a6c58d98aeebf1b23c10c249c813e6e9fb/include/LightGBM/config.h#L283
https://github.com/Microsoft/LightGBM/blob/b9dc51a6c58d98aeebf1b23c10c249c813e6e9fb/include/LightGBM/config.h#L217
https://github.com/Microsoft/LightGBM/blob/b9dc51a6c58d98aeebf1b23c10c249c813e6e9fb/include/LightGBM/config.h#L273

**Comment by guolinke** [comment]
@StrikerRUS 
I am saying the default value of these parameters in sklearn interface.

```seed``` is a overall config for all other seed, it will use this seed to generate many other seeds, e.g. the feature_fraction_seed and so on. refer to : https://github.com/Microsoft/LightGBM/blob/master/src/io/config.cpp#L150-L157  

```top_k``` is used in voting based parallel: https://github.com/Microsoft/LightGBM/blob/master/src/treelearner/voting_parallel_tree_learner.cpp#L15

```machines``` is the machine list used in network: refer to:  https://github.com/Microsoft/LightGBM/blob/master/include/LightGBM/c_api.h#L740

**Comment by StrikerRUS** [comment]
@guolinke As I understand, we don't need to add `seed` to Parameters.rst because it's needed only for sklearn and there is no reason to use it in CLI version. `machines` is created by parsing file path to which is specified in `mlist` parameter. So it seems we need add only description of `top_k` or is it also calculated from other parameters?

To speak about sklearn, `seed` overrides all other "seeds" and they are ignored if user passes them in **kwargs, right?
Other params are connected to network mechanism which isn't implemented in Python yet, yes?

**Comment by guolinke** [comment]
@StrikerRUS 
yeah, I think we only need the ```top_k```, and ```machines``` is designed for the c_api, also not for CLI version.

**Comment by StrikerRUS** [comment]
@guolinke Please formulate the description of `top_k` parameter - I'll add it to docs.

**Comment by guolinke** [comment]
@StrikerRUS 
```top_k```, type int, default 10
used in voting parallel, larger value will bring more accurate result, but slower in training speed.

**Comment by StrikerRUS** [comment]
@guolinke Thank you. But it seems that default is 20:

https://github.com/Microsoft/LightGBM/blob/b9dc51a6c58d98aeebf1b23c10c249c813e6e9fb/include/LightGBM/config.h#L217

**Comment by guolinke** [comment]
@StrikerRUS you are right, sorry for the mistake.

**Comment by StrikerRUS** [comment]
@guolinke no problem, I've added it.

What's the difference between cpp's
```
-  ``bin_construct_sample_cnt``, default=\ ``200000``, type=int

   -  number of data that sampled to construct histogram bins

   -  will give better training result when set this larger, but will increase data loading time

   -  set this to larger value if data is very sparse
```
and sklearn's
`subsample_for_bin (int, optional (default=50000)) – Number of samples for constructing bins.`
?

And the second question: do we need to add a note for all "seeds" about that "for Python-package, this parameter is ignored, use random_state input argument of LGBMModel class instead to set all seeds at once"?

**Comment by StrikerRUS** [comment]
@guolinke Please restart Appveyor and what about this https://github.com/Microsoft/LightGBM/pull/1018#issuecomment-339745184 ?

**Comment by guolinke** [comment]
@StrikerRUS sorry for the late response.
They are the same parameter. 

I am not sure about the second question. 
I think all these seeds are not ignored in c api. Will python package ignore them ? 
@wxchan for the confirm .

**Comment by StrikerRUS** [comment]
>They are the same parameter.

I seems that at present they are not aliases, but independent params:
https://github.com/Microsoft/LightGBM/blob/2b20569e309a295848ea749b3192bec094519f51/include/LightGBM/config.h#L460
https://github.com/Microsoft/LightGBM/blob/2b20569e309a295848ea749b3192bec094519f51/include/LightGBM/config.h#L463
Should I move `subsample_for_bin` to `alias_table` and set its' default value in sklearn to 200000?



About seeds: I suppose here they are ignored in case of `seed` (`random_state` in sklearn wrapper) is passed:
https://github.com/Microsoft/LightGBM/blob/2b20569e309a295848ea749b3192bec094519f51/src/io/config.cpp#L150-L157

**Comment by guolinke** [comment]
@StrikerRUS I think it is a bug, you can fix it.

**Comment by StrikerRUS** [comment]
@guolinke You are right. Seeds are **NOT** ignored even if `seed` is passed. They are overwritten later, I think.

```
import lightgbm as lgb
import numpy as np

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.2, random_state=42)

clf_1 = lgb.sklearn.LGBMClassifier(colsample_bytree=0.8, feature_fraction_seed=1, seed=42)
clf_2 = lgb.sklearn.LGBMClassifier(colsample_bytree=0.8, feature_fraction_seed=2, seed=42)
y_pred_1 = clf_1.fit(X_train, y_train).predict_proba(X_test)
y_pred_2 = clf_2.fit(X_train, y_train).predict_proba(X_test)
np.testing.assert_allclose(y_pred_1, y_pred_2)
```
At least, this code gives error.

**Comment by StrikerRUS** [comment]
@guolinke
> I think it is a bug, you can fix it.

Yeah, at present, it seems that `subsample_for_bin` is totally ignored.

```
import lightgbm as lgb
import numpy as np

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.2, random_state=42)

clf_1 = lgb.sklearn.LGBMClassifier(subsample_for_bin=200000)
clf_2 = lgb.sklearn.LGBMClassifier(subsample_for_bin=5)
y_pred_1 = clf_1.fit(X_train, y_train).predict_proba(X_test)
y_pred_2 = clf_2.fit(X_train, y_train).predict_proba(X_test)
np.testing.assert_allclose(y_pred_1, y_pred_2)
```
This code finishes without error.
```
import lightgbm as lgb
import numpy as np

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split

iris = load_iris()
X_train, X_test, y_train, y_test = train_test_split(iris.data, iris.target, test_size=0.2, random_state=42)

clf_1 = lgb.sklearn.LGBMClassifier(bin_construct_sample_cnt=200000)
clf_2 = lgb.sklearn.LGBMClassifier(bin_construct_sample_cnt=5)
y_pred_1 = clf_1.fit(X_train, y_train).predict_proba(X_test)
y_pred_2 = clf_2.fit(X_train, y_train).predict_proba(X_test)
np.testing.assert_allclose(y_pred_1, y_pred_2)
```
And this one produces the error as it is expected.

**Comment by wxchan** [comment]
Those seeds won't be ignored, it's set after LightGBM/src/io/config.cpp Lines 150 to 157.

**Comment by guolinke** [comment]
@StrikerRUS ready to merge ?
